lib/bolt/project.rb

```diff
@@ -167,5 +167,131 @@ def check_deprecated_file
         Bolt::Logger.deprecation_warning('Using project.yaml instead of bolt-project.yaml', msg)
       end
     end
+
+    def migrate_project(configured_inventory_path, outputter)
+      migrating_inv = configured_inventory_path || inventory_file
+      inventory_1_to_2(migrating_inv, outputter) if migrating_inv.file?
+      bolt_yaml_to_bolt_project(migrating_inv, outputter)
+    end
+
+    def backup_file(origin_path, outputter)
+      unless File.exist?(origin_path)
+        outputter.print_message "Could not find file #{origin_path}, skipping backup."
+        return
+      end
+
+      date = Time.new.strftime("%Y%m%d_%H%M%S%L")
+      backup_dir = File.join(path, '.bolt-bak')
+      FileUtils.mkdir_p(backup_dir)
+
+      filename = File.basename(origin_path)
+      backup_path = File.join(backup_dir, "#{filename}.#{date}.bak")
+
+      outputter.print_message "Backing up #{filename} from #{origin_path} to #{backup_path}"
```

Comment:
Since this is used to backup the inventory in two places, you can get a message that the inventory was backed up twice. I think we should have a flag to tell if a file has already been backed up, or have some other way to not show this message multiple times.

Reply:
Hm, I wonder if giving more context in the message would be useful? Since it's backing up 2 different inventory files, it seems somewhat useful to tell the user we're backing up both times, and so users aren't suprised when there are 2 different invetory backups. Maybe an optional parameter to the method could be a reason for backing up, and the message would be `Backing up inventory.yaml from /home/user/project/inventory.yaml to /home/user/project/.bolt-bak before migrating inventory to version 2` and `Backing up inventory.yaml from /home/user/project/inventory.yaml to /home/user/project/.bolt-bak before moving transport config into inventory.yaml`

Reply:
Since from the user's perspective the project is migrated in a single step, I'm not sure if there's a benefit to having backup files for each step of the migration process. I do think having those specific messages is an improvement, though.

I think ideally the `migrate_project` method would determine which files need to be backed up and then offload the work to actually migrating the files to the other methods... but that could also be difficult considering all the different conditions for determining whether a file needs to be migrated or not.